modules/template/services/TemplateInstanceRenderer.php

```diff
@@ -0,0 +1,51 @@
+<?php
+
+namespace humhub\modules\custom_pages\modules\template\services;
+
+use humhub\modules\custom_pages\lib\templates\TemplateEngineFactory;
+use humhub\modules\custom_pages\modules\template\models\TemplateInstance;
+use Yii;
+use yii\caching\DummyCache;
+
+class TemplateInstanceRendererService
+{
+    public const CACHE_PREFIX = 'cp-';
+
+    private $templateInstance;
+
+    private TemplateService $templateService;
+
+    public function __construct(TemplateInstance $templateInstance)
+    {
+        $this->templateInstance = $templateInstance;
+        $this->templateService = new TemplateService($this->templateInstance->template);
+    }
+
+    public function render()
+    {
+        $cache = ($this->isCachable()) ? Yii::$app->cache : new DummyCache();
+
+        return $cache->getOrSet(static::CACHE_PREFIX . $this->templateInstance->id, function () {
+            $engine = TemplateEngineFactory::create("twig");
+
+            $twigVariables = [];
+            foreach ($this->templateService->getElements() as $element) {
+                $twigVariables[$element->getTemplateName()] = $element->getTemplateValue($this->templateInstance);
+            }
+
+            return $engine->render($this->templateInstance->template->source, $twigVariables);
+        });
+    }
+
+    private function isCachable(): bool
+    {
+        foreach ($this->templateService->getElements() as $element) {
+            if ($element->isDynamic()) {
```

Comment:
Currently we cache per current user if Template has the Element `UserContent` - [here](https://github.com/humhub/custom-pages/commit/7a8302950a5ef714fa854f81b3f91f4a203a1909#diff-566bd82383275914f2b7481beaab3553d4722c6309d6ca640e5f42181849a4ccR142).
Do you want to make `isDynamic() => true` for the user Element?
Or we should implement new method like `getCacheType()` with prossible values:

- `always`|true - Always cache
- `none`|false - Don't cache
- `user` - Cache per current user

Reply:
@yurabakhtin I like the idea of adding different cache levels/types.
However, I'm not sure if we should do that in the early version. There will be some more dynamic elements added. Maybe it makes sense to add more/other cache modes.